core/src/test/java/org/infinispan/functional/FunctionalDistributionTest.java

```diff
@@ -0,0 +1,108 @@
+       * second invocation on the originator happens before the first is
+       * committed so it receives the same input data. Not so in async mode.
+       */
+      isSync = false;
+   }
+
+   @BeforeClass
+   @Override
+   public void createBeforeClass() throws Throwable {
+      super.createBeforeClass();
+   }
+
+   public void testDistributionFromPrimaryOwner() throws InterruptedException {
+      Object key = "testDistributionFromPrimaryOwner";
+      doTestDistribution(key, cacheManagers.stream()
+            .map(cm -> cm.<Object, Integer> getCache(DIST).getAdvancedCache())
+            .filter(cache -> cache.getDistributionManager().getPrimaryLocation(key)
+                  .equals(cache.getRpcManager().getAddress()))
+            .findAny()
+            .get());
+   }
+
+   public void testDistributionFromSecondaryOwner() throws InterruptedException {
+      Object key = "testDistributionFromSecondaryOwner";
+      doTestDistribution(key, cacheManagers.stream()
+            .map(cm -> cm.<Object, Integer> getCache(DIST).getAdvancedCache())
+            // owner...
+            .filter(cache -> cache.getDistributionManager().getLocality(key).isLocal()
+                  // ...but not primary owner
+                  && !cache.getDistributionManager().getPrimaryLocation(key)
+                        .equals(cache.getRpcManager().getAddress()))
+            .findAny()
+            .get());
+   }
+
+   public void testDistributionFromNonOwner() throws InterruptedException {
+      Object key = "testDistributionFromNonOwner";
+      doTestDistribution(key, cacheManagers.stream()
+            .map(cm -> cm.<Object, Integer> getCache(DIST).getAdvancedCache())
+            .filter(cache -> !cache.getDistributionManager().getLocality(key).isLocal())
+            .findAny()
+            .get());
+   }
+
+   private void doTestDistribution(Object key, AdvancedCache<Object, Integer> originator) throws InterruptedException {
+      ReadWriteMap<Object, Integer> rw = ReadWriteMapImpl.create(FunctionalMapImpl.create(originator));
+
+      // with empty cache:
+      iterate(key, rw, 1);
+      // again:
+      iterate(key, rw, 2);
+   }
+
+   private void iterate(Object key, ReadWriteMap<Object, Integer> rw, int expectedValue) throws InterruptedException {
+      rw.eval(key, (Function<ReadWriteEntryView<Object, Integer>, Void> & Serializable)
+            entry -> {
+               // we need a small delay so that the value gets committed before the replication finishes:
+               try {
+                  Thread.sleep(10);
+               } catch (InterruptedException e) {
+                  Thread.currentThread().interrupt();
+               }
+               return entry.set(entry.find().orElse(0) + 1);
+            }).join();
+
+      // since we're in async mode, we need to wait before the replication finishes
+      Thread.sleep(100); // TODO: find a better way?
```

Comment:
Minor thing, TestingUtil.sleepThread() sleeps and does not throw InterruptedException.